{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tissue-specific genes based on MGI Gene Expression Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import mean dNdS scores for each mouse protein coding gene, indexed with Ensembl ID. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "info_df = pd.read_csv('../data/info.mouse_protein_coding_genes.tsv',sep='\\t',header=0,index_col=0)\n",
    "info_df.drop_duplicates(subset='Gene name',inplace=True) # Drop the duplicated gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_df = pd.read_csv('../results/Ensembl98_mouse/mouse.dNdS_stats.all_genes.tsv',sep='\\t',index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import MGI IDs for genes specific to each organ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "MGI_df = pd.read_csv('../results/MGI_organs/MGI_ID-tissue_type.tsv',sep='\\t',index_col=0,names=['MGI ID','Tissue'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import the matching ensembl IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ensembl_df = pd.read_csv('../data/MGI_organs/Ensembl_ID-MGI_ID.tsv',sep='\\t',index_col=1)\n",
    "Ensembl_df = Ensembl_df[~Ensembl_df.index.duplicated(keep='first')] # remove duplicated Ensembl IDs for the same gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_df = pd.merge(MGI_df,Ensembl_df, left_index=True, right_index=True, how='right')\n",
    "tissue_df = tissue_df.reset_index().set_index('Gene stable ID')\n",
    "tissue_df = pd.merge(tissue_df,stats_df,left_index=True,right_index=True,how='left')\n",
    "tissue_df.dropna(inplace=True)\n",
    "tissue_df=pd.merge(info_df,tissue_df,left_index=True,right_index=True,how='right')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>MGI ID</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ENSMUSG00000000049</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>apolipoprotein H [Source:MGI Symbol;Acc:MGI:88...</td>\n",
       "      <td>MGI:88058</td>\n",
       "      <td>liver</td>\n",
       "      <td>74.0</td>\n",
       "      <td>0.215338</td>\n",
       "      <td>0.047188</td>\n",
       "      <td>0.125469</td>\n",
       "      <td>0.186157</td>\n",
       "      <td>0.204020</td>\n",
       "      <td>0.224057</td>\n",
       "      <td>0.389129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSMUSG00000000049</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>apolipoprotein H [Source:MGI Symbol;Acc:MGI:88...</td>\n",
       "      <td>MGI:88058</td>\n",
       "      <td>liver</td>\n",
       "      <td>74.0</td>\n",
       "      <td>0.215338</td>\n",
       "      <td>0.047188</td>\n",
       "      <td>0.125469</td>\n",
       "      <td>0.186157</td>\n",
       "      <td>0.204020</td>\n",
       "      <td>0.224057</td>\n",
       "      <td>0.389129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSMUSG00000000605</td>\n",
       "      <td>Clcn4</td>\n",
       "      <td>chloride channel, voltage-sensitive 4 [Source:...</td>\n",
       "      <td>MGI:104571</td>\n",
       "      <td>heart</td>\n",
       "      <td>77.0</td>\n",
       "      <td>0.016013</td>\n",
       "      <td>0.025864</td>\n",
       "      <td>0.005139</td>\n",
       "      <td>0.008509</td>\n",
       "      <td>0.009622</td>\n",
       "      <td>0.011225</td>\n",
       "      <td>0.189252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSMUSG00000000605</td>\n",
       "      <td>Clcn4</td>\n",
       "      <td>chloride channel, voltage-sensitive 4 [Source:...</td>\n",
       "      <td>MGI:104571</td>\n",
       "      <td>heart</td>\n",
       "      <td>77.0</td>\n",
       "      <td>0.016013</td>\n",
       "      <td>0.025864</td>\n",
       "      <td>0.005139</td>\n",
       "      <td>0.008509</td>\n",
       "      <td>0.009622</td>\n",
       "      <td>0.011225</td>\n",
       "      <td>0.189252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSMUSG00000000794</td>\n",
       "      <td>Kcnn3</td>\n",
       "      <td>potassium intermediate/small conductance calci...</td>\n",
       "      <td>MGI:2153183</td>\n",
       "      <td>heart</td>\n",
       "      <td>77.0</td>\n",
       "      <td>0.047890</td>\n",
       "      <td>0.032956</td>\n",
       "      <td>0.012382</td>\n",
       "      <td>0.037018</td>\n",
       "      <td>0.043321</td>\n",
       "      <td>0.050230</td>\n",
       "      <td>0.280899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSMUSG00000109336</td>\n",
       "      <td>Samd4b</td>\n",
       "      <td>sterile alpha motif domain containing 4B [Sour...</td>\n",
       "      <td>MGI:2448542</td>\n",
       "      <td>brain</td>\n",
       "      <td>79.0</td>\n",
       "      <td>0.049798</td>\n",
       "      <td>0.039833</td>\n",
       "      <td>0.017568</td>\n",
       "      <td>0.031868</td>\n",
       "      <td>0.039254</td>\n",
       "      <td>0.051114</td>\n",
       "      <td>0.272936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSMUSG00000110086</td>\n",
       "      <td>Gm45623</td>\n",
       "      <td>predicted gene 45623 [Source:MGI Symbol;Acc:MG...</td>\n",
       "      <td>MGI:5791459</td>\n",
       "      <td>brain</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.053854</td>\n",
       "      <td>0.025075</td>\n",
       "      <td>0.012989</td>\n",
       "      <td>0.044657</td>\n",
       "      <td>0.050247</td>\n",
       "      <td>0.062072</td>\n",
       "      <td>0.161806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSMUSG00000110086</td>\n",
       "      <td>Gm45623</td>\n",
       "      <td>predicted gene 45623 [Source:MGI Symbol;Acc:MG...</td>\n",
       "      <td>MGI:5791459</td>\n",
       "      <td>brain</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.053854</td>\n",
       "      <td>0.025075</td>\n",
       "      <td>0.012989</td>\n",
       "      <td>0.044657</td>\n",
       "      <td>0.050247</td>\n",
       "      <td>0.062072</td>\n",
       "      <td>0.161806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSMUSG00000112129</td>\n",
       "      <td>Pbld1</td>\n",
       "      <td>phenazine biosynthesis-like protein domain con...</td>\n",
       "      <td>MGI:1915621</td>\n",
       "      <td>kidney</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.204431</td>\n",
       "      <td>0.062927</td>\n",
       "      <td>0.144982</td>\n",
       "      <td>0.168154</td>\n",
       "      <td>0.174836</td>\n",
       "      <td>0.242931</td>\n",
       "      <td>0.314268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSMUSG00000112129</td>\n",
       "      <td>Pbld1</td>\n",
       "      <td>phenazine biosynthesis-like protein domain con...</td>\n",
       "      <td>MGI:1915621</td>\n",
       "      <td>kidney</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.204431</td>\n",
       "      <td>0.062927</td>\n",
       "      <td>0.144982</td>\n",
       "      <td>0.168154</td>\n",
       "      <td>0.174836</td>\n",
       "      <td>0.242931</td>\n",
       "      <td>0.314268</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1405 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene name  \\\n",
       "Gene stable ID                 \n",
       "ENSMUSG00000000049      Apoh   \n",
       "ENSMUSG00000000049      Apoh   \n",
       "ENSMUSG00000000605     Clcn4   \n",
       "ENSMUSG00000000605     Clcn4   \n",
       "ENSMUSG00000000794     Kcnn3   \n",
       "...                      ...   \n",
       "ENSMUSG00000109336    Samd4b   \n",
       "ENSMUSG00000110086   Gm45623   \n",
       "ENSMUSG00000110086   Gm45623   \n",
       "ENSMUSG00000112129     Pbld1   \n",
       "ENSMUSG00000112129     Pbld1   \n",
       "\n",
       "                                                     Gene description  \\\n",
       "Gene stable ID                                                          \n",
       "ENSMUSG00000000049  apolipoprotein H [Source:MGI Symbol;Acc:MGI:88...   \n",
       "ENSMUSG00000000049  apolipoprotein H [Source:MGI Symbol;Acc:MGI:88...   \n",
       "ENSMUSG00000000605  chloride channel, voltage-sensitive 4 [Source:...   \n",
       "ENSMUSG00000000605  chloride channel, voltage-sensitive 4 [Source:...   \n",
       "ENSMUSG00000000794  potassium intermediate/small conductance calci...   \n",
       "...                                                               ...   \n",
       "ENSMUSG00000109336  sterile alpha motif domain containing 4B [Sour...   \n",
       "ENSMUSG00000110086  predicted gene 45623 [Source:MGI Symbol;Acc:MG...   \n",
       "ENSMUSG00000110086  predicted gene 45623 [Source:MGI Symbol;Acc:MG...   \n",
       "ENSMUSG00000112129  phenazine biosynthesis-like protein domain con...   \n",
       "ENSMUSG00000112129  phenazine biosynthesis-like protein domain con...   \n",
       "\n",
       "                         MGI ID  Tissue  count      mean       std       min  \\\n",
       "Gene stable ID                                                                 \n",
       "ENSMUSG00000000049    MGI:88058   liver   74.0  0.215338  0.047188  0.125469   \n",
       "ENSMUSG00000000049    MGI:88058   liver   74.0  0.215338  0.047188  0.125469   \n",
       "ENSMUSG00000000605   MGI:104571   heart   77.0  0.016013  0.025864  0.005139   \n",
       "ENSMUSG00000000605   MGI:104571   heart   77.0  0.016013  0.025864  0.005139   \n",
       "ENSMUSG00000000794  MGI:2153183   heart   77.0  0.047890  0.032956  0.012382   \n",
       "...                         ...     ...    ...       ...       ...       ...   \n",
       "ENSMUSG00000109336  MGI:2448542   brain   79.0  0.049798  0.039833  0.017568   \n",
       "ENSMUSG00000110086  MGI:5791459   brain   50.0  0.053854  0.025075  0.012989   \n",
       "ENSMUSG00000110086  MGI:5791459   brain   50.0  0.053854  0.025075  0.012989   \n",
       "ENSMUSG00000112129  MGI:1915621  kidney    9.0  0.204431  0.062927  0.144982   \n",
       "ENSMUSG00000112129  MGI:1915621  kidney    9.0  0.204431  0.062927  0.144982   \n",
       "\n",
       "                         25%       50%       75%       max  \n",
       "Gene stable ID                                              \n",
       "ENSMUSG00000000049  0.186157  0.204020  0.224057  0.389129  \n",
       "ENSMUSG00000000049  0.186157  0.204020  0.224057  0.389129  \n",
       "ENSMUSG00000000605  0.008509  0.009622  0.011225  0.189252  \n",
       "ENSMUSG00000000605  0.008509  0.009622  0.011225  0.189252  \n",
       "ENSMUSG00000000794  0.037018  0.043321  0.050230  0.280899  \n",
       "...                      ...       ...       ...       ...  \n",
       "ENSMUSG00000109336  0.031868  0.039254  0.051114  0.272936  \n",
       "ENSMUSG00000110086  0.044657  0.050247  0.062072  0.161806  \n",
       "ENSMUSG00000110086  0.044657  0.050247  0.062072  0.161806  \n",
       "ENSMUSG00000112129  0.168154  0.174836  0.242931  0.314268  \n",
       "ENSMUSG00000112129  0.168154  0.174836  0.242931  0.314268  \n",
       "\n",
       "[1405 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tissue_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "brain_df = tissue_df[tissue_df['Tissue']=='brain']\n",
    "brain_arr = brain_df['mean'].dropna().values\n",
    "heart_df = tissue_df[tissue_df['Tissue']=='heart']\n",
    "heart_arr = heart_df['mean'].dropna().values\n",
    "kidney_df = tissue_df[tissue_df['Tissue']=='kidney']\n",
    "kidney_arr = kidney_df['mean'].dropna().values\n",
    "liver_df = tissue_df[tissue_df['Tissue']=='liver']\n",
    "liver_arr = liver_df['mean'].dropna().values\n",
    "lung_df = tissue_df[tissue_df['Tissue']=='lung']\n",
    "lung_arr = lung_df['mean'].dropna().values\n",
    "pancreas_df = tissue_df[tissue_df['Tissue']=='pancreas']\n",
    "pancreas_arr = pancreas_df['mean'].dropna().values\n",
    "skin_df = tissue_df[tissue_df['Tissue']=='skin']\n",
    "skin_arr = skin_df['mean'].dropna().values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import statsmodels.api as sm\n",
    "import seaborn as sns\n",
    "import scipy.stats as stats\n",
    "import matplotlib.ticker as mtick"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "matplotlib.rcParams['figure.dpi']= 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n",
      "The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n"
     ]
    }
   ],
   "source": [
    "# Creating a figure \n",
    "fig = plt.figure(figsize=(10,7.5)) # Size of a letter size paper in horizontal\n",
    "fig.suptitle('Distribution of dN/dS of Mouse Tissue-specific Genes', fontsize=14)\n",
    "\n",
    "# Setting subplot space\n",
    "grid = plt.GridSpec(nrows=1,ncols=1)\n",
    "#grid.update(wspace=0.5, hspace=0.3)\n",
    "\n",
    "# The subplot for distribution histogram \n",
    "distr_plot = fig.add_subplot(grid[:,:])\n",
    "#distr_hist = dist_GO_df.plot(ax=distr_plot, kind='hist',alpha=0.3,bins=np.logspace(np.log10(0.001),np.log10(10), 100), logx=True, color=['royalblue','orange','lightgreen'])\n",
    "\n",
    "# Set up the bins for log scale x-axis, and get the centers\n",
    "bins=np.logspace(np.log10(0.001),np.log10(10), 50)\n",
    "bins_cntr = (bins[1:] + bins[:-1]) / 2\n",
    "\n",
    "# brain-specific genes' dN/dS histogram and curve\n",
    "brain_counts, brain_bin_edges, ignored = distr_plot.hist(brain_arr,bins,histtype='stepfilled',alpha=0.3,weights=np.ones(len(brain_arr)) / len(brain_arr),color='cyan',label='dN/dS of brain-specific genes (med={0:.3f})'.format(np.median(brain_arr)))\n",
    "try:\n",
    "    # calculate area of histograms (area under PDF should be 1)\n",
    "    brain_area_hist = ((brain_bin_edges[1:] - brain_bin_edges[:-1]) * brain_counts).sum()\n",
    "    brain_shape, brain_loc, brain_scale = stats.lognorm.fit(brain_arr)\n",
    "    # pdf-values using cdf \n",
    "    brain_fit_log_cntr_ = stats.lognorm.cdf(bins, brain_shape, loc=brain_loc, scale=brain_scale)\n",
    "    brain_fit_log_cntr = np.diff(brain_fit_log_cntr_)\n",
    "    # plot fitted and scaled PDFs into histogram\n",
    "    distr_plot.plot(bins_cntr, brain_fit_log_cntr * brain_counts.sum(),'c-', \n",
    "                    label='lognormal fit of brain-specific dN/dS distribution', linewidth=2)\n",
    "except ValueError:\n",
    "    pass\n",
    "\n",
    "# heart-specific genes' dN/dS histogram and curve\n",
    "heart_counts, heart_bin_edges, ignored = distr_plot.hist(heart_arr,bins,histtype='stepfilled',alpha=0.3,weights=np.ones(len(heart_arr)) / len(heart_arr),color='red',label='dN/dS of heart-specific genes (med={0:.3f})'.format(np.median(heart_arr)))\n",
    "try:\n",
    "    # calculate area of histograms (area under PDF should be 1)\n",
    "    heart_area_hist = ((heart_bin_edges[1:] - heart_bin_edges[:-1]) * heart_counts).sum()\n",
    "    heart_shape, heart_loc, heart_scale = stats.lognorm.fit(heart_arr)\n",
    "    # pdf-values using cdf \n",
    "    heart_fit_log_cntr_ = stats.lognorm.cdf(bins, heart_shape, loc=heart_loc, scale=heart_scale)\n",
    "    heart_fit_log_cntr = np.diff(heart_fit_log_cntr_)\n",
    "    # plot fitted and scaled PDFs into histogram\n",
    "    distr_plot.plot(bins_cntr, heart_fit_log_cntr * heart_counts.sum(),'r-', \n",
    "                    label='lognormal fit of heart-specific dN/dS distribution', linewidth=2)\n",
    "except ValueError:\n",
    "    pass\n",
    "\n",
    "# kidney-specific genes' dN/dS histogram and curve\n",
    "kidney_counts, kidney_bin_edges, ignored = distr_plot.hist(kidney_arr,bins,histtype='stepfilled',alpha=0.3,weights=np.ones(len(kidney_arr)) / len(kidney_arr),color='black',label='dN/dS of kidney-specific genes (med={0:.3f})'.format(np.median(kidney_arr)))\n",
    "try:\n",
    "    # calculate area of histograms (area under PDF should be 1)\n",
    "    kidney_area_hist = ((kidney_bin_edges[1:] - kidney_bin_edges[:-1]) * kidney_counts).sum()\n",
    "    kidney_shape, kidney_loc, kidney_scale = stats.lognorm.fit(kidney_arr)\n",
    "    # pdf-values using cdf \n",
    "    kidney_fit_log_cntr_ = stats.lognorm.cdf(bins, kidney_shape, loc=kidney_loc, scale=kidney_scale)\n",
    "    kidney_fit_log_cntr = np.diff(kidney_fit_log_cntr_)\n",
    "    # plot fitted and scaled PDFs into histogram\n",
    "    distr_plot.plot(bins_cntr, kidney_fit_log_cntr * kidney_counts.sum(),'k-', \n",
    "                    label='lognormal fit of kidney-specific dN/dS distribution', linewidth=2)\n",
    "except ValueError:\n",
    "    pass\n",
    "\n",
    "# liver-specific genes' dN/dS histogram and curve\n",
    "liver_counts, liver_bin_edges, ignored = distr_plot.hist(liver_arr,bins,histtype='stepfilled',alpha=0.3,weights=np.ones(len(liver_arr)) / len(liver_arr),color='green',label='dN/dS of liver-specific genes (med={0:.3f})'.format(np.median(liver_arr)))\n",
    "try:\n",
    "    # calculate area of histograms (area under PDF should be 1)\n",
    "    liver_area_hist = ((liver_bin_edges[1:] - liver_bin_edges[:-1]) * liver_counts).sum()\n",
    "    liver_shape, liver_loc, liver_scale = stats.lognorm.fit(liver_arr)\n",
    "    # pdf-values using cdf \n",
    "    liver_fit_log_cntr_ = stats.lognorm.cdf(bins, liver_shape, loc=liver_loc, scale=liver_scale)\n",
    "    liver_fit_log_cntr = np.diff(liver_fit_log_cntr_)\n",
    "    # plot fitted and scaled PDFs into histogram\n",
    "    distr_plot.plot(bins_cntr, liver_fit_log_cntr * liver_counts.sum(),'g-', \n",
    "                    label='lognormal fit of liver-specific dN/dS distribution', linewidth=2)\n",
    "except ValueError:\n",
    "    pass\n",
    "\n",
    "# lung-specific genes' dN/dS histogram and curve\n",
    "lung_counts, lung_bin_edges, ignored = distr_plot.hist(lung_arr,bins,histtype='stepfilled',alpha=0.3,weights=np.ones(len(lung_arr)) / len(lung_arr),color='magenta',label='dN/dS of lung-specific genes (med={0:.3f})'.format(np.median(lung_arr)))\n",
    "try:\n",
    "    # calculate area of histograms (area under PDF should be 1)\n",
    "    lung_area_hist = ((lung_bin_edges[1:] - lung_bin_edges[:-1]) * lung_counts).sum()\n",
    "    lung_shape, lung_loc, lung_scale = stats.lognorm.fit(lung_arr)\n",
    "    # pdf-values using cdf \n",
    "    lung_fit_log_cntr_ = stats.lognorm.cdf(bins, lung_shape, loc=lung_loc, scale=lung_scale)\n",
    "    lung_fit_log_cntr = np.diff(lung_fit_log_cntr_)\n",
    "    # plot fitted and scaled PDFs into histogram\n",
    "    distr_plot.plot(bins_cntr, lung_fit_log_cntr * lung_counts.sum(),'m-', \n",
    "                    label='lognormal fit of lung-specific dN/dS distribution', linewidth=2)\n",
    "except ValueError:\n",
    "    pass\n",
    "\n",
    "# pancreas-specific genes' dN/dS histogram and curve\n",
    "pancreas_counts, pancreas_bin_edges, ignored = distr_plot.hist(pancreas_arr,bins,histtype='stepfilled',alpha=0.3,weights=np.ones(len(pancreas_arr)) / len(pancreas_arr),color='blue',label='dN/dS of pancreas-specific genes (med={0:.3f})'.format(np.median(pancreas_arr)))\n",
    "try:\n",
    "    # calculate area of histograms (area under PDF should be 1)\n",
    "    pancreas_area_hist = ((pancreas_bin_edges[1:] - pancreas_bin_edges[:-1]) * pancreas_counts).sum()\n",
    "    pancreas_shape, pancreas_loc, pancreas_scale = stats.lognorm.fit(pancreas_arr)\n",
    "    # pdf-values using cdf \n",
    "    pancreas_fit_log_cntr_ = stats.lognorm.cdf(bins, pancreas_shape, loc=pancreas_loc, scale=pancreas_scale)\n",
    "    pancreas_fit_log_cntr = np.diff(pancreas_fit_log_cntr_)\n",
    "    # plot fitted and scaled PDFs into histogram\n",
    "    distr_plot.plot(bins_cntr, pancreas_fit_log_cntr * pancreas_counts.sum(),'b-', \n",
    "                    label='lognormal fit of pancreas-specific dN/dS distribution', linewidth=2)\n",
    "except ValueError:\n",
    "    pass\n",
    "\n",
    "# skin-specific genes' dN/dS histogram and curve\n",
    "skin_counts, skin_bin_edges, ignored = distr_plot.hist(skin_arr,bins,histtype='stepfilled',alpha=0.3,weights=np.ones(len(skin_arr)) / len(skin_arr),color='yellow',label='dN/dS of skin-specific genes (med={0:.3f})'.format(np.median(skin_arr)))\n",
    "try:\n",
    "    # calculate area of histograms (area under PDF should be 1)\n",
    "    skin_area_hist = ((skin_bin_edges[1:] - skin_bin_edges[:-1]) * skin_counts).sum()\n",
    "    skin_shape, skin_loc, skin_scale = stats.lognorm.fit(skin_arr)\n",
    "    # pdf-values using cdf \n",
    "    skin_fit_log_cntr_ = stats.lognorm.cdf(bins, skin_shape, loc=skin_loc, scale=skin_scale)\n",
    "    skin_fit_log_cntr = np.diff(skin_fit_log_cntr_)\n",
    "    # plot fitted and scaled PDFs into histogram\n",
    "    distr_plot.plot(bins_cntr, skin_fit_log_cntr * skin_counts.sum(),'y-', \n",
    "                    label='lognormal fit of skin-specific dN/dS distribution', linewidth=2)\n",
    "except ValueError:\n",
    "    pass\n",
    "\n",
    "# Axis labels\n",
    "distr_plot.set_xlabel(xlabel='dN/dS')\n",
    "distr_plot.set_ylabel(ylabel='percentage of genes')\n",
    "distr_plot.set_xscale('log')\n",
    "distr_plot.legend(loc='best')\n",
    "distr_plot.yaxis.set_major_formatter(mtick.PercentFormatter(1))\n",
    "\n",
    "fig.savefig('../figures/mouse.organs.pdf')\n",
    "fig.savefig('../figures/mouse.organs.eps')\n",
    "fig.savefig('../figures/mouse.organs.png')\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
